Docs: Plugin E2E - CI article (#777)
grafana/plugin-tools: docusaurus/docs/e2e-test-a-plugin/ci.md

---
id: ci
title: CI Workflow
description: How to run end-to-end tests in CI
keywords:
  - grafana
  - plugins
  - plugin
  - e2e
  - end-to-end
  - ci
sidebar_position: 30
---

import BEPluginWorkflowNPM from '@snippets/plugin-e2e-ds-workflow.npm.md';
import BEPluginWorkflowYarn from '@snippets/plugin-e2e-ds-workflow.yarn.md';
import BEPluginWorkflowPNPM from '@snippets/plugin-e2e-ds-workflow.pnpm.md';
import FEPluginNPM from '@snippets/plugin-e2e-fe-plugin-workflow.npm.md';
import FEPluginYarn from '@snippets/plugin-e2e-fe-plugin-workflow.yarn.md';
import FEPluginPNPM from '@snippets/plugin-e2e-fe-plugin-workflow.pnpm.md';

## Introduction

> We recommend using a CI workflow to run end-to-end tests to continuously check for breakages. The following Github workflows will run end-to-end tests against a range of Grafana versions for every PR in your Github repository.

Author: Looks good, but removed the `improve code quality` bit.

:::

<details>
  <summary> <h3>Backend plugin workflow</h3> </summary>
  <CodeSnippets
snippets={[
{ component: BEPluginWorkflowNPM, label: 'npm' },
{ component: BEPluginWorkflowYarn, label: 'yarn' },
{ component: BEPluginWorkflowPNPM, label: 'pnpm' }
]}
groupId="package-manager"
queryString="current-package-manager"
/>
</details>

<details>
  <summary> <h3>Frontend plugin workflow</h3> </summary>
  <CodeSnippets
snippets={[
{ component: BEPluginWorkflowNPM, label: 'npm' },
{ component: BEPluginWorkflowYarn, label: 'yarn' },
{ component: BEPluginWorkflowPNPM, label: 'pnpm' }
]}
groupId="package-manager"
queryString="current-package-manager"
/>
</details>

### The e2e-versions Action

The example workflows have a job called `Resolve Grafana images` that uses the [e2e-version](https://github.com/grafana/plugin-actions/tree/main/e2e-version) Action to resolve a list of Grafana images. For every image returned, a new job will be fired up that builds the plugin, starts Grafana and runs the end-to-end tests.

The Action supports two modes - `plugin-grafana-dependency` and `version-support-policy`.

> #### plugin-grafana-dependency mode

Author: Looks good, but as said in another PR, I'll open a separate PR that moves to use code formatting so we have consistent headers across all plugin-e2e docs.

> By default, the e2e-version Action will return the most recent grafana-dev image plus all the latest patch releases of every minor version of Grafana Enterprise since the version that was specified as `grafanaDependency` in the [plugin.json](../metadata.md). To avoid starting too many jobs, to output will be capped 6 versions.

Reviewer: ```suggestion
By default, the e2e-version Action returns the most recent grafana-dev image. Additionally, the Action returns all the latest patch releases of most minor versions of Grafana Enterprise. Specifically, it returns the minor versions of Grafana Enterprise since the version that was specified as `grafanaDependency` in the [plugin.json](../metadata.md). To avoid starting too many jobs, the output is capped at 6 versions.
```

Reviewer: Docs style recommends against sentences longer than 25 words. In this case, I needed to split the original sentence into 3 separate sentences to enhance readability. Please check that it is still accurate.

Author: Not entirely correct. Tried to shorten the text. Let me know what you think.

> By default, the e2e-version Action will return the most recent grafana-dev image plus all the latest patch releases of every minor version of Grafana Enterprise since the version that was specified as `grafanaDependency` in the [plugin.json](../metadata.md). To avoid starting too many jobs, to output will be capped 6 versions.

Reviewer 2: Could we add some "smartness" here return versions depending on how the `grafanaDependency` is specified.

If it is specified like `grafanaDependency: ^9.8.1` we only return the latest patch version of the `9.8` release.

If it is specified `grafanaDependency: >=9.8.1` it would be as you described above etc.

> By default, the e2e-version Action will return the most recent grafana-dev image plus all the latest patch releases of every minor version of Grafana Enterprise since the version that was specified as `grafanaDependency` in the [plugin.json](../metadata.md). To avoid starting too many jobs, to output will be capped 6 versions.

Author: That would be nice! Think in general, this Action needs to have more flexibility. For example when we run the tests in CI for the plugin-e2e repo, it uses the `version-support-policy`. That means we're only testing from 9.5, although we have committed to support every version from Grafana 8.5. So would be nice if the Action had mode that allowed you to specify a min version, and also skip the capping of versions. 

I'm making a note on your proposal. 👍

![](/img/e2e-version-plugin-dependency.png)

#### version-support-policy mode

> Except for resolving the most recent grafana-dev image, the `version-support-policy` mode will resolve versions according to Grafana's plugin compatibility support policy. Specifically, it retrieves the latest patch release for each minor version within the current major version of Grafana. Additionally, it includes the most recent release for the latest minor version of the previous major Grafana version.

Author: Is the `version-support-policy` official @sympatheticmoose? Would be nice to link to some docs here.

Reviewer 2: Yeah, with the current documentation it doesn't describe what that policy is. Which would be good.

> Except for resolving the most recent grafana-dev image, the `version-support-policy` mode will resolve versions according to Grafana's plugin compatibility support policy. Specifically, it retrieves the latest patch release for each minor version within the current major version of Grafana. Additionally, it includes the most recent release for the latest minor version of the previous major Grafana version.

Reviewer: ```suggestion
Except for resolving the most recent `grafana-dev` image, the `version-support-policy` mode resolves versions according to Grafana's plugin compatibility support policy. Specifically, it retrieves the latest patch release for each minor version within the current major version of Grafana. Additionally, it includes the most recent release for the latest minor version of the previous major Grafana version.
```

![](/img/e2e-version-version-support-policy.png)

> To use the `version-support-policy` mode you need to specify the `version-resolver-type` input argument.

Reviewer: ```suggestion
To use the `version-support-policy` mode, you need to specify the `version-resolver-type` input argument like in this example:
```

```yml
- name: Resolve Grafana E2E versions
        id: resolve-versions
        uses: grafana/plugin-actions/e2e-version
        with:
          version-resolver-type: version-support-policy
```

### Playwright report

> A Playwright HTML test report is generated for every Grafana version that is being tested. In case any of the tests failed, a Playwright trace viewer is also generated along with the report. The `Upload artifacts` step in the example workflows uploads the report as artifacts. To find information on how to download and view the report, refer to the Playwright [documentatation](https://playwright.dev/docs/ci-intro#html-report).

Reviewer: ```suggestion
A Playwright HTML test report is generated for every Grafana version that is being tested. In case any of the tests fail, a Playwright trace viewer is also generated along with the report. The `Upload artifacts` step in the example workflows uploads the report as an artifact. To find information on how to download and view the report, refer to the [Playwright documentatation](https://playwright.dev/docs/ci-intro#html-report).
```
